### PR TITLE
memory usage monitor

### PR DESCRIPTION
**Pull request by ymatsuda**
@eishay 
MemoryUsageMonitor will send an airbrake notification when two GCs in a raw were unable to reduce the heap usage less than 80%. This only checks the CMS old gen heap.

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/common/healthcheck/MemoryUsageMonitor.scala [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import java.lang.management.MemoryNotificationInfo
+import java.lang.management.MemoryPoolMXBean
+import java.lang.management.MemoryType
+import javax.management.Notification
+import javax.management.NotificationListener
+import javax.management.NotificationEmitter
+import scala.collection.JavaConversions._
+import com.keepit.common.logging.Logging
+
+object MemoryUsageMonitor {
```
Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/common/healthcheck/MemoryUsageMonitor.scala:14]: I would avoid having objects as factories. instead have a guice module that constructs the object at hand (as with `AirbrakeModule` etc).

File: kifi-backend/modules/common/app/com/keepit/FortyTwoGlobal.scala [hunk 2]
```diff
@@ -70,8 +70,12 @@ abstract class FortyTwoGlobal(val mode: Mode.Mode)
       log.error("STARTUP SELF CHECK FAILED!")
     }
     serviceDiscovery.forceUpdate()
+
+    memoryUsageMonitor = MemoryUsageMonitor(injector.instance[AirbrakeNotifier])
   }
 
+  private var memoryUsageMonitor: MemoryUsageMonitor = null
```
Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/FortyTwoGlobal.scala]: Why would we need that reference, is it used?

File: kifi-backend/modules/common/app/com/keepit/common/healthcheck/MemoryUsageMonitor.scala [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        case MemoryNotificationInfo.MEMORY_COLLECTION_THRESHOLD_EXCEEDED =>
+          collectionUsageThresholdCount += 1
+          warn(memPool.bean, memPool.threshold, memPool.maxHeapSize, collectionUsageThresholdCount)
+        case _ =>
+      }
+    }
+  }
+
+  val emitter = ManagementFactory.getMemoryMXBean.asInstanceOf[NotificationEmitter]
+  monitoredPools.foreach{ memPool => emitter.addNotificationListener(listener, null, memPool) }
```
Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/common/healthcheck/MemoryUsageMonitor.scala]: Is simple creation of the class means that it starts listening?
Wouldn't you want to have a `startListen()` method?

**Comment by eishay** [comment]
great idea!

**Comment by ymatsuda** [comment]
@eishay  MemoryUsageMonitor is now injected.

**Comment by eishay** [comment]
great, thanks!

**Pull request merged**